src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -111,74 +154,164 @@ internal abstract class AbstractAddParameterCodeFixProvider<
+                var signature = $"{method.Name}({string.Join(", ", parameters)})";
+
+                var title = string.Format(FeaturesResources.Add_parameter_to_0, signature);
+
+                context.RegisterCodeFix(
+                    new MyCodeAction(title, c => FixAsync(context.Document, method, argumentToAdd, arguments, typeArgument, c)),
+                    context.Diagnostics);
+            }
+        }
+
+        private ImmutableArray<(IMethodSymbol method, TArgumentSyntax argument, int index, TTypeSyntax typeArgument)> GetArgumentInsertPositionAndTypeArgumentForMethodCandidates(
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFacts,
+            ImmutableArray<IMethodSymbol> methodCandidates,
+            SeparatedSyntaxList<TArgumentSyntax> arguments,
+            TArgumentSyntax argumentOpt,
+            SeparatedSyntaxList<TTypeSyntax> typeArguments)
+        {
+            var methodsAndArgumentToAdd = ArrayBuilder<(IMethodSymbol constructor, TArgumentSyntax argument, int index, TTypeSyntax typeArgument)>.GetInstance();
 
-            foreach (var constructor in type.InstanceConstructors.OrderBy(m => m.Parameters.Length))
+            foreach (var method in methodCandidates.OrderBy(m => m.Parameters.Length))
             {
-                if (constructor.IsNonImplicitAndFromSource() &&
-                    NonParamsParameterCount(constructor) < arguments.Count)
+                if (method.IsNonImplicitAndFromSource())
                 {
-                    var argumentToAdd = DetermineFirstArgumentToAdd(
-                        semanticModel, syntaxFacts, comparer, constructor,
-                        arguments, argumentOpt);
+                    var argumentToAdd = GetArgumentInsertPositionForMethodCandidate(semanticModel, syntaxFacts, method, arguments, argumentOpt);
+                    var typeArgumentToAdd = GetTypeArgumentForMethodCandidate(method, typeArguments);
 
-                    if (argumentToAdd != null)
+                    if (argumentToAdd != null || typeArgumentToAdd != null)
                     {
-                        if (argumentOpt != null && argumentToAdd != argumentOpt)
-                        {
-                            // We were trying to fix a specific argument, but the argument we want
-                            // to fix is something different.  That means there was an error earlier
-                            // than this argument.  Which means we're looking at a non-viable 
-                            // constructor.  Skip this one.
-                            continue;
-                        }
-
-                        constructorsAndArgumentToAdd.Add(
-                            (constructor, argumentToAdd, arguments.IndexOf(argumentToAdd)));
+                        methodsAndArgumentToAdd.Add(
+                            (method, argumentToAdd, arguments.IndexOf(argumentToAdd), typeArgumentToAdd));
```

Comment:
a named type would make it easy to also explain that "argumentToInsert" can be null, and "argumentInsertionIndex" can be negative.